### PR TITLE
enhancement(redis source): Add redis streams source support (#15335)

### PR DESCRIPTION
**Pull request by IdoKendo**

### CODE REVIEW

**Comment by jszwedko** [comment]
Just leaving a note for reviewers that a previous PR adding stream support to sinks, https://github.com/vectordotdev/vector/pull/10172, didn't make it in because streams were different enough from the other supported data types (lists and channels) that it seemed better to have a separate sink for it (`redis_streams`). I'm not sure if it is the same for the source, but just something to consider.

File: src/sources/redis/stream.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        result_string.push(' ');
+        result_string.push_str(redis_value_to_string(value).as_str());
+        result_string.push(' ');
+    }
+
+    result_string
+}
+
+async fn read_stream(mut conn: Connection, key: String) -> RedisResult<String> {
+    let opts = StreamReadOptions::default().count(1).block(0);
```
Comment (StephenWakely) [inline src/sources/redis/stream.rs:64]: We don't need to read just one event at a time. This should read as many events as necessary and create a `Vec<Event>` to send to the output.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Hi, many thanks for putting this together!

As alluded to by @jszwedko,  Redis streams doesn't really fit into the data model that the current source provides. This needs to be a separate source - call it `redis_streams`.

Each event returned by the stream is already a map, each field of that map should be a field in the `Event` that is created. It doesn't make sense to include a `decoder` option in the configuration since essentially the `StreamId` has already been decoded.

There might be some code that can be shared between this source and the new one, but I think it should be pretty minimal, so it is probably fine to just have it duplicated.

Let me know if you have any questions or issues doing this.

**Comment by jszwedko** [comment]
Given the lack of activity here and the fact that we'd want to take a different approach of adding a new `redis_streams` source I think we can close this out. Thanks for getting the ball rolling on this idea though @IdoKendo ! It can serve as a reference for anyone else that picks up this work.

**Pull request closed without merging**